### PR TITLE
Use Wasm memory.size instruction to get the wasm heap size

### PR DESCRIPTION
**Pull request by juj**
Discovered the Wasm `memory.size` instruction while searching for something else. This PR changes to use that instruction to get the wasm heap size instead of jumping out to JS. Fix includes of fiber.h and heap.h and update the test.

This PR assumes that Wasm `memory.size` instruction is ubiquitous and was available already in Wasm MVP. I don't remember if that is the case. Tried to find out from the spec history if that was so, but ended up with https://github.com/WebAssembly/design/issues/1395 . @kripken @dschuff do you recall if `memory.size` instruction has always been there?

### CODE REVIEW

**Comment by dschuff** [comment]
Memory.size has been in wasm since the beginning, so I'm pretty sure this should be fine. Does it need to be exported (I guess not, a user could just export it manually via command line or whatever?)

**Review by kripken — APPROVED** [review]
Yes, this looks good to me too.

The header changes in other files could perhaps be split out, but I don't feel strongly. Maybe explain that in the commit message though.

**Comment by haberbyte** [comment]
Quick question to this:

Would that mean the below line would be the same as calling emscripten_get_heap_size, therefore the entire #ifdef can be removed?

https://github.com/emscripten-core/emscripten/blob/9bb322f8a7ee89d6ac67e828b9c7a7022ddf8de2/system/lib/sbrk.c#L75

**Comment by juj** [comment]
@haberbyte Yes, updated the code.

@kripken It looks like tests are failing in wasm2js. Maybe wasm2js conversion does not implement the `memory.size` opcode? That is probably why that existing `__wasm__` preprocessor directive was present in `sbrk()`?

**Comment by kripken** [comment]
@juj 

No, wasm2js implements them,

https://github.com/WebAssembly/binaryen/blob/56acee228d93d8ca36e40e4eb659e91d4a058f6f/src/wasm2js.h#L1969-L1971

(we were missing a test I see, I'm adding that in https://github.com/WebAssembly/binaryen/pull/3547)

Note that the `__wasm__` directive in the source code is true for wasm and for wasm2js. wasm2js only occurs during link, after we've compiled to wasm normally, we then compile the wasm to JS.

(I don't see wasm2js failures though - it's erroring on the build stage?)

**Comment by juj** [comment]
> (I don't see wasm2js failures though - it's erroring on the build stage?)

Thanks, yeah, looks like I misinterpreted. Revised the test to drop the now redundant JS dep, but now test runner is giving an internal LLVM error:

```
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/libc++-except.a... (this will be cached in "/root/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++-except.a" for subsequent builds)
clang++: /b/s/w/ir/cache/builder/emscripten-releases/llvm-project/llvm/include/llvm/ADT/SmallVector.h:295: llvm::SmallVectorTemplateCommon::reference llvm::SmallVectorTemplateCommon<const llvm::MachineBasicBlock *>::back() [T = const llvm::MachineBasicBlock *]: Assertion `!empty()' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /root/emsdk/upstream/bin/clang++ -DEMSCRIPTEN -fno-inline-functions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Xclang -iwithsysroot/include/SDL -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=13 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration --sysroot=/root/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -Werror -DLIBCXX_BUILDING_LIBCXXABI=1 -D_LIBCPP_BUILDING_LIBRARY -Oz -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -fwasm-exceptions -c /root/project/system/lib/libcxx/src/condition_variable.cpp -o /root/emsdk/upstream/emscripten/cache/build/libc++-except/condition_variable.o
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/root/project/system/lib/libcxx/src/condition_variable.cpp'.
4.	Running pass 'WebAssembly CFG Stackify' on function '@_ZNSt3__225notify_all_at_thread_exitERNS_18condition_variableENS_11unique_lockINS_5mutexEEE'
 #0 0x00007f0254e4e173 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x1864173)
 #1 0x00007f0254e4bebe llvm::sys::RunSignalHandlers() (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x1861ebe)
 #2 0x00007f0254e4d4ed llvm::sys::CleanupOnSignal(unsigned long) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x18634ed)
 #3 0x00007f0254d84c43 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x179ac43)
 #4 0x00007f0254d84dae CrashRecoverySignalHandler(int) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x179adae)
 #5 0x00007f02533dd890 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12890)
 #6 0x00007f024f6ace97 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3ee97)
 #7 0x00007f024f6ae801 abort (/lib/x86_64-linux-gnu/libc.so.6+0x40801)
 #8 0x00007f024f69e39a (/lib/x86_64-linux-gnu/libc.so.6+0x3039a)
 #9 0x00007f024f69e412 (/lib/x86_64-linux-gnu/libc.so.6+0x30412)
#10 0x00007f0256d9d193 (anonymous namespace)::WebAssemblyCFGStackify::placeMarkers(llvm::MachineFunction&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x37b3193)
#11 0x00007f0256d96063 (anonymous namespace)::WebAssemblyCFGStackify::runOnMachineFunction(llvm::MachineFunction&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x37ac063)
#12 0x00007f025523bc1d llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x1c51c1d)
#13 0x00007f0254fbb218 llvm::FPPassManager::runOnFunction(llvm::Function&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x19d1218)
#14 0x00007f0254fc1cf1 llvm::FPPassManager::runOnModule(llvm::Module&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x19d7cf1)
#15 0x00007f0254fbb8bc llvm::legacy::PassManagerImpl::run(llvm::Module&) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x19d18bc)
#16 0x00007f02521e5fa5 (anonymous namespace)::EmitAssemblyHelper::EmitAssemblyWithNewPassManager(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x21d0fa5)
#17 0x00007f02521e0612 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x21cb612)
#18 0x00007f025250760c clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x24f260c)
#19 0x00007f02512e6644 clang::ParseAST(clang::Sema&, bool, bool) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x12d1644)
#20 0x00007f0252a7d830 clang::FrontendAction::Execute() (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x2a68830)
#21 0x00007f0252a000ea clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x29eb0ea)
#22 0x00007f0252aefcb8 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x2adacb8)
#23 0x0000000000214ba9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/root/emsdk/upstream/bin/clang+++0x214ba9)
#24 0x0000000000212a72 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) (/root/emsdk/upstream/bin/clang+++0x212a72)
#25 0x00007f025270f982 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::$_1>(long) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x26fa982)
#26 0x00007f0254d84b57 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/root/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x179ab57)
#27 0x00007f025270f4d7 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x26fa4d7)
#28 0x00007f02526dfb6f clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x26cab6f)
#29 0x00007f02526dfe67 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x26cae67)
#30 0x00007f02526f760c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/root/emsdk/upstream/bin/../lib/libclang-cpp.so.13git+0x26e260c)
#31 0x00000000002123cd main (/root/emsdk/upstream/bin/clang+++0x2123cd)
#32 0x00007f024f68fb97 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b97)
#33 0x000000000020f87a _start (/root/emsdk/upstream/bin/clang+++0x20f87a)
clang-13: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 0fc1738eb75d613b9e16143b83e7cb80512e84eb)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /root/emsdk/upstream/bin
clang-13: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-13: note: diagnostic msg: /tmp/condition_variable-ffadeb.cpp
clang-13: note: diagnostic msg: /tmp/condition_variable-ffadeb.sh
clang-13: note: diagnostic msg: 
```

**Comment by juj** [comment]
Looks like the internal error occurs independent of this PR, it also happens on this PR #13444 .

**Comment by juj** [comment]
Also the other errors on test-mac also reproduce on latest master, so do not seem to be related to this issue. I think this should be safe to land, unless people want to get the master branch green first.

**Comment by kripken** [comment]
The main branch is green now, so we can see if all tests pass here with a merge.

**Comment by bdenhollander** [comment]
@kripken This PR didn't auto-merge.

**Comment by kripken** [comment]
Looks like test failures prevented it, that look like non-flakes.

File: system/lib/libc/emscripten_get_heap_size.c [hunk 1]
```diff
@@ -5,8 +5,8 @@
  * found in the LICENSE file.
  */
 
-#include <emscripten/heap.h>
```
Comment (sbc100) [inline system/lib/libc/emscripten_get_heap_size.c:8]: Why this change?   Normally we want to include the header file where the function we are implement is declared no?

File: system/lib/libc/emscripten_get_heap_size.c [hunk 2]
```diff
@@ -5,8 +5,8 @@
...
+#include <stddef.h>
 
 size_t emscripten_get_heap_size() {
-  return __builtin_wasm_memory_size(0) * WASM_PAGE_SIZE;
+  return __builtin_wasm_memory_size(0) << 16;
```
Comment (sbc100) [inline system/lib/libc/emscripten_get_heap_size.c:11]: Won't to optimizer take care of doing this?

Comment (juj) [reply system/lib/libc/emscripten_get_heap_size.c:11]: I don't know, but it certainly won't change a << 16 back to a * 65536. I don't think it is that good to rely on optimizer to achieve what one wants, but better to just write out what one wants directly when it is trivially possible.

File: system/lib/libc/emscripten_get_heap_size.c (hunk 1 above)
Comment (juj) [reply system/lib/libc/emscripten_get_heap_size.c:8]: I don't recall why this change, it does look accidental now.
